core/src/main/java/org/pentaho/di/core/database/DatabaseFactory.java

```diff
@@ -0,0 +1,154 @@
+/*! ******************************************************************************
+ *
+ * Pentaho Data Integration
+ *
+ * Copyright (C) 2002-2018 by Hitachi Vantara : http://www.pentaho.com
+ *
+ *******************************************************************************
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************************/
+
+package org.pentaho.di.core.database;
+
+import org.pentaho.di.core.Const;
+import org.pentaho.di.core.exception.KettleDatabaseException;
+import org.pentaho.di.core.exception.KettleException;
+import org.pentaho.di.core.logging.LoggingObjectInterface;
+import org.pentaho.di.core.logging.LoggingObjectType;
+import org.pentaho.di.core.logging.SimpleLoggingObject;
+import org.pentaho.di.i18n.BaseMessages;
+
+/**
+ *
+ * @author matt
+ *
+ */
+public class DatabaseFactory implements DatabaseFactoryInterface {
+
+  private static Class<?> PKG = Database.class; // for i18n purposes, needed by Translator2!!
+  private boolean success;
+  private Exception capturedException;
+
+  public static final LoggingObjectInterface loggingObject = new SimpleLoggingObject(
+    "Database factory", LoggingObjectType.GENERAL, null );
+
+  public DatabaseFactory() {
+  }
+
+  @Override
+  public String getConnectionTestReport( DatabaseMeta databaseMeta ) {
+    success = true; // default
+    if ( databaseMeta.getAccessType() != DatabaseMeta.TYPE_ACCESS_PLUGIN ) {
+      StringBuilder report = new StringBuilder();
+
+      Database db = new Database( loggingObject, databaseMeta );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *db*. The resource is referenced at lines 61, 78, 79, 85, 100, 102. The resource is closed at lines 75, 97. There are other execution paths that do not contain closure statements, for example, when an if condition at line 57 evaluates to true.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).